x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -495,6 +512,49 @@ public boolean maybeFetchFullEntry(KeyType cacheKey, long length, RangeMissingHa
         return true;
     }
 
+    /**
+     * Fetch and write in cache a region of a blob if there are enough free pages in the cache to do so.
+     *
+     * This method returns as soon as the download tasks are instantiated, but the tasks themselves
+     * are run on the bulk executor.
+     *
+     * If an exception is thrown from the writer then the cache entry being downloaded is freed
+     * and unlinked
+     *
+     * @param cacheKey  the key to fetch data for
+     * @param region    the region of the blob to fetch
+     * @param blobLength the length of the blob from which the region is fetched (used to compute the size of the ending region)
+     * @param writer    a writer that handles writing of newly downloaded data to the shared cache
+     * @param listener  listener that is called once all downloading has finished
+     */
+    public void maybeFetchRegion(
+        final KeyType cacheKey,
+        final int region,
+        final long blobLength,
+        final RangeMissingHandler writer,
+        final ActionListener<Void> listener
+    ) {
+        if (freeRegionCount() < 1) {
```

Comment:
I wonder if we should consider both free regions and regions at level 0 as free here?

I believe we'll never eagerly evict data just because it is unused. So here we might have data in cache that has not been used for many hours but still do no eager fetching.

Reply:
Yes, I think it makes sense to also consider regions that are infrequently used, as long as they spent a bit of time in the cache to avoid evicting newly created cache entries.

I gave it a try in the latest changes by introducing a `maybeEvictLeastUsed` that takes both frequency and last access time. Let me know what you think.